zaza/openstack/charm_tests/neutron/tests.py

```diff
@@ -293,8 +293,23 @@ def setUpClass(cls):
         cls.neutron_client = (
             openstack_utils.get_neutron_session_client(cls.keystone_session))
 
+    def _test_400_additional_validation(self, expected_network_names):
+        """Additional assertions for test_400_create_network.
+
+        Can be overridden in derived classes.
+
+        :type expected_network_names: List[str]
+        """
+        pass
+
     def test_400_create_network(self):
-        """Create a network, verify that it exists, and then delete it."""
+        """Create a network, verify that it exists, and then delete it.
```

Comment:
This feels a bit awkward, but I can't quite put my finger on it.  What I'm referring to is the introduction of `self._test_400_additional_validation()` at three points in this test, as a form of extending the test and allowing overrides.

I'll try to explain:

 * It seems to be specializing the `_test_400_additional_validation()` using parameter, but it's not really clear how it's supposed to be used.
 * With this kind of adding functionality would normally be done with a `super()` call, but obviously there are 3 points of change.

It's not very helpful for me to just say: "Well, I don't like that for 'reasons'"!  So I'll try and explain what I might do differently: I think there are two options:

1. I'd split the contents of `test_400_create_network()` into 3 sub function helpers, and call those from the test.  Then either those functions can be 'wrapped' (overridden) or `test_400_create_network` can be overridden, and insert the helper validation as well as calling the 3 sub functions.
2. As `test_400_create_network()` is small, I'd probably just override the whole function in the using test, and just add in the bits that are needed; yes, there's code duplication, but there's not a lot, and it's obvious.  However, if bugs are found, they have to be fixed in two places.